endpoint/OBDComp/emulator/obd_emulator.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "common/ta_errors.h"
+#include "endpoint/OBDComp/can-bus/can-utils.h"
+#include "endpoint/OBDComp/obd_pid.h"
+#include "obd_emulator.h"
+
+int service_01_response(int fd, int pid);
+
+void print_help(void) { printf("Usage: obd_emulator <can_interface>\n"); }
+
+int main(int argc, char* argv[]) {
+  char* default_can_device = OBD_EMULATOR_INTERFACE;
+  if (argc >= 2 && strncmp(argv[1], "-h", strlen("-h")) == 0) {
+    print_help();
+    exit(EXIT_SUCCESS);
+  }
+
+  if (argc >= 2 && argv[1] != NULL) {
+    default_can_device = argv[1];
+  }
+
+  int s;
+  status_t ret;
+  ret = can_open(default_can_device, &s);
+  if (ret != SC_OK) {
+    fprintf(stderr, "%s\n", ta_error_to_string(ret));
+    exit(EXIT_FAILURE);
+  }
+
+  printf("OBDII emulator start!\n");
+
+  struct can_frame frame;
+  do {
+    ret = can_recv(s, &frame);
+    if (ret != SC_OK) {
+      fprintf(stderr, "%s\n", ta_error_to_string(ret));
+      continue;
+    }
+    printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
+    for (int i = 0; i < frame.can_dlc; i++) printf("%02X ", frame.data[i]);
+    printf("\r\n");
+
+    if (frame.can_id == 0x7DF) {
+      printf("Received OBD query\r\n");
+      if (frame.can_dlc >= 2) {
+        switch (frame.data[1]) {
+          case 1:
+            printf("Service 01: Show current data\r\n");
+            service_01_response(s, frame.data[2]);
+            break;
+          default:
+            printf("Unknown service 0x%02X", frame.data[1]);
+            break;
+        }
+      } else {
+        printf("Error, frame too short. DLC = %d bytes.\r\n", frame.can_dlc);
```

Comment:
If we encountered problems, shall we continue trying? Or, wait for certain seconds for next query?

Reply:
I think it is better to wait for the next query. This error indicates that the query data is invalid.